src/coreComponents/common/LifoStorageCuda.hpp

```diff
@@ -0,0 +1,215 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+#ifndef LIFOSTORAGECUDA_HPP
+#define LIFOSTORAGECUDA_HPP
+
+#include <deque>
+#include <future>
+#include <mutex>
+#include <condition_variable>
+#include <camp/camp.hpp>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <algorithm>
+
+#include "common/FixedSizeDeque.hpp"
+#include "common/GEOS_RAJA_Interface.hpp"
+#include "common/TimingMacros.hpp"
+#include "common/FixedSizeDequeWithMutexes.hpp"
+#include "common/MultiMutexesLock.hpp"
+#include "common/LifoStorageCommon.hpp"
+
+namespace geosx
+{
+/**
+ * This class is used to store in a LIFO way buffers, first on device, then on host, then on disk.
+ */
+
+template< typename T, typename INDEX_TYPE >
+class LifoStorageCuda : public LifoStorageCommon< T, INDEX_TYPE >
+{
+  typedef LifoStorageCommon< T, INDEX_TYPE > baseLifo;
+public:
+
+
+  /**
+   * A LIFO storage will store numberOfBuffersToStoreDevice buffer on
+   * deevice, numberOfBuffersToStoreHost on host and the rest on disk.
+   *
+   * @param name                           Prefix of the files used to save the occurenncy of the saved buffer on disk.
+   * @param elemCnt                        Number of elments in the LvArray we want to store in the LIFO storage.
+   * @param numberOfBuffersToStoreOnDevice Maximum number of array to store on device memory ( -1 = use 80% of remaining memory ).
+   * @param numberOfBuffersToStoreOnHost   Maximum number of array to store on host memory ( -1 = use 80% of remaining memory ).
+   * @param maxNumberOfBuffers             Number of arrays expected to be stores in the LIFO.
+   */
+  LifoStorageCuda( std::string name, size_t elemCnt, int numberOfBuffersToStoreOnDevice, int numberOfBuffersToStoreOnHost, int maxNumberOfBuffers ):
+    LifoStorageCommon< T, INDEX_TYPE >( name, elemCnt, numberOfBuffersToStoreOnHost, maxNumberOfBuffers ),
+    m_deviceDeque( numberOfBuffersToStoreOnDevice, elemCnt, LvArray::MemorySpace::cuda ),
+    m_pushToDeviceEvents( maxNumberOfBuffers ),
+    m_popFromDeviceEvents( maxNumberOfBuffers )
+  {}
+
+  /**
+   * Asynchroneously push a copy of the given LvArray into the LIFO
+   *
+   * @param array The LvArray to store in the LIFO, should match the size of the data used in constructor.
+   */
+  void pushAsync( arrayView1d< T > array )
```

Comment:
And w.r.t. to this pattern, do not forget to add the `override` keyword, and most probably the `final` as well (maybe define the whole class as `final` will be easier.)
That way, if you directly use the derived class and not the base class, the compiler will be able to make some optimisations.

Reply:
Done